src/HonzaBotner.Services/CvutAuthorizationService.cs

```diff
@@ -40,42 +40,64 @@ public class CvutAuthorizationService : IAuthorizationService
             _logger = logger;
         }
 
-        public async Task<bool> AuthorizeAsync(string accessToken, string username, ulong userId, RolesPool rolesPool)
+        public async Task<IAuthorizationService.AuthorizeResult> AuthorizeAsync(string accessToken, string username, ulong userId, RolesPool rolesPool)
         {
-            bool requiresUnverified = rolesPool == RolesPool.Auth;
-
-            if (await IsUserVerified(userId) && requiresUnverified)
-            {
-                _logger.LogWarning("User already verified");
-                return false;
-            }
+            bool discordIdPresent = await IsUserVerified(userId);
 
             UsermapPerson? person = await _usermapInfoService.GetUserInfoAsync(accessToken, username);
             if (person == null)
             {
                 _logger.LogWarning("Couldn't fetch info from UserMap");
-                return false;
+                return IAuthorizationService.AuthorizeResult.UserMapError;
             }
 
             string authId = _hashService.Hash(person.Username);
-            if (await _dbContext.Verifications.AnyAsync(v => v.AuthId == authId) && requiresUnverified)
-            {
-                _logger.LogWarning("User already verified 2");
-                return false;
-            }
+            bool authPresent = await _dbContext.Verifications.AnyAsync(v => v.AuthId == authId);
 
             IReadOnlySet<DiscordRole> discordRoles = _roleManager.MapUsermapRoles(person.Roles, rolesPool);
-            bool rolesGranted = await _roleManager.GrantRolesAsync(userId, discordRoles);
 
-            if (rolesGranted && requiresUnverified)
+            // discord and auth -> update roles
+            if (discordIdPresent && authPresent)
             {
-                Verification verification = new() {AuthId = authId, UserId = userId};
+                Verification discordVerification = await _dbContext.Verifications.FirstAsync(v => v.UserId == userId);
+                Verification authVerification = await _dbContext.Verifications.FirstAsync(v => v.AuthId == authId);
```

Comment:
We can use Any to look into database. So maybe use something like `Verifications.Any(v => v.AuthId == authId && v.UserId == userId)`